prime-router/src/main/kotlin/fhirengine/engine/FHIRTranslator.kt

```diff
@@ -47,83 +50,124 @@ class FHIRTranslator(
+     * Accepts a [FhirTranslateQueueMessage] [message] and, based on its parameters, sends a report to the next pipeline
+     * step containing either the first ancestor's blob or a new blob that has been translated per
+     * the receiver's settings, pending the passed topic's (found in [message]) isSendOriginal property
+     * [actionHistory] and [actionLogger] ensure all activities are recorded to the database and logged
      */
     override fun <T : QueueMessage> doWork(
         message: T,
         actionLogger: ActionLogger,
         actionHistory: ActionHistory,
     ): List<FHIREngineRunResult> {
         message as ReportPipelineMessage
-
-        logger.trace("Translating FHIR file for receivers.")
-        // pull fhir document and parse FHIR document
-        val bundle = FhirTranscoder.decode(message.downloadContent())
-
-        // track input report
-        actionHistory.trackExistingInputReport(message.reportId)
-
-        when (message) {
-            is FhirTranslateQueueMessage -> {
-                val receiver = settings.findReceiver(message.receiverFullName)
-                    ?: throw RuntimeException("Receiver with name ${message.receiverFullName} was not found")
-                actionHistory.trackActionReceiverInfo(receiver.organizationName, receiver.name)
-
-                var nextAction = Event.EventAction.BATCH
-                var externalName: String? = null
-                var queueMessage: ReportEventQueueMessage? = null
-                val bodyBytes = if (message.topic.isSendOriginal) {
-                    nextAction = Event.EventAction.SEND
-                    val originalReport = getOriginalReport(message.reportId)
-                    externalName = originalReport.externalName
-                    queueMessage = ReportEventQueueMessage(nextAction, false, message.reportId, "")
-                    BlobAccess.downloadBlobAsByteArray(originalReport.bodyUrl)
-                } else {
-                    getByteArrayFromBundle(receiver, bundle)
-                }
-
-                // get a Report from the message
-                val (report, event, blobInfo) = Report.generateReportAndUploadBlob(
-                    nextAction,
-                    bodyBytes,
-                    listOf(message.reportId),
-                    receiver,
-                    this.metadata,
-                    actionHistory,
-                    topic = message.topic,
-                    externalName
-                )
-
-                if (queueMessage != null) {
-                    queueMessage = queueMessage.copy(reportId = report.id)
+        val contextMap = mapOf(
+            MDCUtils.MDCProperty.ACTION_NAME to actionHistory.action.actionName.name,
+            MDCUtils.MDCProperty.REPORT_ID to message.reportId,
+            MDCUtils.MDCProperty.TOPIC to message.topic,
+            MDCUtils.MDCProperty.BLOB_URL to message.blobURL
+        )
+        withLoggingContext(contextMap) {
```

Comment:
This is now a sergeant method that calls the two worker methods. This was done to better separate the functionality and create maintainability and readability.